### PR TITLE
Fix shape errors in CompartmentalModel

### PR DESCRIPTION
**Pull request by fritzo**
I still haven't been able to locally reproduce the CI errors. However this PR:
- adds more stringent assertions, and changes shaping logic to conform
- adds more unit tests, and fixes errors caught by unit tests
- adds more verbose error messages to previously flaky assertions, so they are easier to diagnose in case of future failure

For reference, here's an example of the error I haven't been able to reproduce:

<details>

```
2020-06-17T02:47:41.8016294Z =================================== FAILURES ===================================
2020-06-17T02:47:41.8018118Z ______________________ test_simple_sir_smoke[svi-{}-0-3] _______________________
2020-06-17T02:47:41.8018250Z 
2020-06-17T02:47:41.8020910Z duration = 3, forecast = 0, options = {}, algo = 'svi'
2020-06-17T02:47:41.8021331Z 
2020-06-17T02:47:41.8021761Z     @pytest.mark.filterwarnings("ignore:num_chains")
2020-06-17T02:47:41.8022130Z     @pytest.mark.parametrize("duration", [3, 7])
2020-06-17T02:47:41.8022502Z     @pytest.mark.parametrize("forecast", [0, 7])
2020-06-17T02:47:41.8022897Z     @pytest.mark.parametrize("algo,options", [
2020-06-17T02:47:41.8023276Z         ("svi", {}),
2020-06-17T02:47:41.8023653Z         ("svi", {"haar": False}),
2020-06-17T02:47:41.8024006Z         ("mcmc", {}),
2020-06-17T02:47:41.8024353Z         ("mcmc", {"haar": True}),
2020-06-17T02:47:41.8024739Z         ("mcmc", {"haar_full_mass": 2}),
2020-06-17T02:47:41.8025093Z         ("mcmc", {"num_quant_bins": 2}),
2020-06-17T02:47:41.8025443Z         ("mcmc", {"num_quant_bins": 4}),
2020-06-17T02:47:41.8025797Z         ("mcmc", {"num_quant_bins": 8}),
2020-06-17T02:47:41.8026146Z         ("mcmc", {"num_quant_bins": 12}),
2020-06-17T02:47:41.8026698Z         ("mcmc", {"num_quant_bins": 16}),
2020-06-17T02:47:41.8027190Z         ("mcmc", {"num_quant_bins": 2, "haar": True}),
2020-06-17T02:47:41.8027564Z         ("mcmc", {"arrowhead_mass": True}),
2020-06-17T02:47:41.8027922Z         ("mcmc", {"jit_compile": True}),
2020-06-17T02:47:41.8028285Z         ("mcmc", {"jit_compile": True, "haar_full_mass": 2}),
2020-06-17T02:47:41.8028668Z         ("mcmc", {"jit_compile": True, "num_quant_bins": 2}),
2020-06-17T02:47:41.8029247Z         ("mcmc", {"num_chains": 2, "mp_context": "spawn"}),
2020-06-17T02:47:41.8029632Z         ("mcmc", {"num_chains": 2, "mp_context": "spawn", "num_quant_bins": 2}),
2020-06-17T02:47:41.8030980Z         ("mcmc", {"num_chains": 2, "mp_context": "spawn", "jit_compile": True}),
2020-06-17T02:47:41.8031389Z     ], ids=str)
2020-06-17T02:47:41.8031747Z     def test_simple_sir_smoke(duration, forecast, options, algo):
2020-06-17T02:47:41.8032099Z         population = 100
2020-06-17T02:47:41.8032439Z         recovery_time = 7.0
2020-06-17T02:47:41.8032771Z     
2020-06-17T02:47:41.8033110Z         # Generate data.
2020-06-17T02:47:41.8033516Z         model = SimpleSIRModel(population, recovery_time, [None] * duration)
2020-06-17T02:47:41.8033900Z         assert model.full_mass == [("R0", "rho")]
2020-06-17T02:47:41.8034243Z         for attempt in range(100):
2020-06-17T02:47:41.8034597Z             data = model.generate({"R0": 1.5, "rho": 0.5})["obs"]
2020-06-17T02:47:41.8034965Z             if data.sum():
2020-06-17T02:47:41.8035304Z                 break
2020-06-17T02:47:41.8035647Z         assert data.sum() > 0, "failed to generate positive data"
2020-06-17T02:47:41.8036042Z     
2020-06-17T02:47:41.8036381Z         # Infer.
2020-06-17T02:47:41.8036729Z         model = SimpleSIRModel(population, recovery_time, data)
2020-06-17T02:47:41.8037065Z         num_samples = 5
2020-06-17T02:47:41.8037404Z         if algo == "mcmc":
2020-06-17T02:47:41.8037771Z             model.fit_mcmc(warmup_steps=1, num_samples=num_samples, max_tree_depth=2, **options)
2020-06-17T02:47:41.8038134Z         else:
2020-06-17T02:47:41.8038480Z             model.fit_svi(num_steps=2, num_samples=num_samples, **options)
2020-06-17T02:47:41.8038824Z     
2020-06-17T02:47:41.8039158Z         # Predict and forecast.
2020-06-17T02:47:41.8039504Z >       samples = model.predict(forecast=forecast)
2020-06-17T02:47:41.8039781Z 
2020-06-17T02:47:41.8040140Z tests/contrib/epidemiology/test_models.py:65: 
2020-06-17T02:47:41.8040502Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-06-17T02:47:41.8041262Z /opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/site-packages/torch/autograd/grad_mode.py:15: in decorate_context
2020-06-17T02:47:41.8041762Z     return func(*args, **kwargs)
2020-06-17T02:47:41.8042127Z /opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/contextlib.py:30: in inner
2020-06-17T02:47:41.8042482Z     return func(*args, **kwds)
2020-06-17T02:47:41.8042845Z /opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/contextlib.py:30: in inner
2020-06-17T02:47:41.8043200Z     return func(*args, **kwds)
2020-06-17T02:47:41.8043544Z pyro/contrib/epidemiology/compartmental.py:546: in predict
2020-06-17T02:47:41.8043897Z     trace = poutine.trace(model).get_trace()
2020-06-17T02:47:41.8044239Z pyro/poutine/trace_messenger.py:187: in get_trace
2020-06-17T02:47:41.8044578Z     self(*args, **kwargs)
2020-06-17T02:47:41.8044918Z pyro/poutine/trace_messenger.py:165: in __call__
2020-06-17T02:47:41.8045257Z     ret = self.fn(*args, **kwargs)
2020-06-17T02:47:41.8045594Z pyro/poutine/messenger.py:11: in _context_wrap
2020-06-17T02:47:41.8045934Z     return fn(*args, **kwargs)
2020-06-17T02:47:41.8046271Z pyro/poutine/messenger.py:11: in _context_wrap
2020-06-17T02:47:41.8046605Z     return fn(*args, **kwargs)
2020-06-17T02:47:41.8046990Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-06-17T02:47:41.8047293Z 
2020-06-17T02:47:41.8047754Z self = <pyro.contrib.epidemiology.models.SimpleSIRModel object at 0x7f7eac92e550>
2020-06-17T02:47:41.8048131Z 
2020-06-17T02:47:41.8048471Z     def _sequential_model(self):
2020-06-17T02:47:41.8048807Z         """
2020-06-17T02:47:41.8049153Z         Sequential model used to sample latents in the interval [0:duration].
2020-06-17T02:47:41.8049512Z         This is compatible with both quantized and relaxed inference.
2020-06-17T02:47:41.8049875Z         This method is called only inside particle_plate.
2020-06-17T02:47:41.8050386Z         """
2020-06-17T02:47:41.8050715Z         C = len(self.compartments)
2020-06-17T02:47:41.8051053Z         T = self.duration
2020-06-17T02:47:41.8051401Z         R_shape = getattr(self.population, "shape", ())  # Region shape.
2020-06-17T02:47:41.8051753Z         num_samples = len(next(iter(self.samples.values())))
2020-06-17T02:47:41.8052092Z     
2020-06-17T02:47:41.8052434Z         # Sample global parameters and auxiliary variables.
2020-06-17T02:47:41.8052781Z         params = self.global_model()
2020-06-17T02:47:41.8053180Z         auxiliary, non_compartmental = self._sample_auxiliary()
2020-06-17T02:47:41.8053531Z     
2020-06-17T02:47:41.8053866Z         # Reshape to accommodate the time_plate below.
2020-06-17T02:47:41.8054211Z >       assert auxiliary.shape == (num_samples, C, T) + R_shape, auxiliary.shape
2020-06-17T02:47:41.8054561Z E       AssertionError: torch.Size([5, 2, 3])
2020-06-17T02:47:41.8055433Z 
2020-06-17T02:47:41.8055932Z pyro/contrib/epidemiology/compartmental.py:780: AssertionError
```

</details>

## Tested
- [x] triggered travis tests 3 times, saw no failure.

### CODE REVIEW

File: pyro/contrib/epidemiology/compartmental.py [hunk 1]
```diff
@@ -390,9 +391,13 @@ def fit_svi(self, *,
             model_trace = poutine.trace(
                 poutine.replay(particle_plate(model), guide_trace)).get_trace()
             self.samples = {name: site["value"] for name, site in model_trace.nodes.items()
-                            if site["type"] == "sample"}
+                            if site["type"] == "sample"
```
Comment (neerajprad) [inline pyro/contrib/epidemiology/compartmental.py:394]: Interesting, I didn't know we could have multiple `if`s in dict/list comprehensions. 🙂

**Comment by neerajprad** [comment]
I'll merge and test this on the github actions PR.

File: pyro/contrib/epidemiology/compartmental.py (hunk 1 above)
Comment (fritzo) [reply pyro/contrib/epidemiology/compartmental.py:394]: Yeah, it's a nice way to break up a long `and` condition.

**Comment by fritzo** [comment]
Thanks for reviewing!

**Pull request merged**